python/ray/data/_internal/plan.py

```diff
@@ -449,16 +424,16 @@ def meta_count(self) -> Optional[int]:
         Returns:
             The number of records of the result Dataset, or None.
         """
-        if self.needs_eager_execution():
-            # If the plan is input/read only, we execute it, so snapshot has output.
-            # This applies to newly created dataset. For example, initial dataset
-            # from read, and output datasets of Dataset.split().
-            self.execute()
-
-        if self._logical_plan.dag is self._snapshot_operator:
-            return self._get_num_rows_from_blocks_metadata(self._snapshot_blocks)
+        if (
```

Comment:
The eager execution check was relevant in the case where the plan was from in-memory only or read-only. In this case, it ensured that `_snapshot_blocks` was not `None` so that we can get metadata from it.

Since we can now get the metadata from in-memory or read operator directly, we don't need to call `execute`.